src/rl/dqn.py

```diff
@@ -0,0 +1,277 @@
+        })
+
+    def _sync_target_network(self):
+        self.sess.run(self.target_update_op)
+
+    def train(self, policy, max_samples):
+        """
+        Train the DQN Agent
+
+        Parameters
+        ---------
+        env: gym.core.Env
+            gym environment to train on
+        policy: rl.policy.Policy
+            the action selection policy to use
+        num_samples: int
+            the number of samples to train for
+
+        returns: void
+        """
+
+        init_op = tf.global_variables_initializer()
+
+        saver = tf.train.Saver(max_to_keep=None)
+
+        with tf.Session() as sess:
+            self.sess = sess
+            checkpoint = tf.train.latest_checkpoint(self.save_dir)
+            if (checkpoint is not None):
+                saver.restore(sess, checkpoint)
+
+            sess.run(init_op)
+
+            n_samples = 0
+            n_updates = 0
+            n_episodes = 0
+            episode_reward = 0  # keep track of this to see if we are learning
+            while (n_samples <= max_samples + self.num_burn_in):
+                state = self.env.reset()
+                is_terminal = False
+                while (not is_terminal):
+                    action = self._select_action(state, policy)
+                    next_state, reward, is_terminal, _ = self.env.step(action)
+                    episode_reward += reward
+                    self.memory.append(Sample(
+                        cur_state=state,
+                        action=action,
+                        next_state=next_state,
+                        reward=reward,
+                        is_terminal=is_terminal
+                    ))
+
+                    if (n_samples > self.num_burn_in):
+                        if (n_samples % self.train_interval == 0):
+                            batch = self.memory.sample(self.batch_size)
+                            (b_s, b_a, b_ns, b_r, b_t) = \
+                                process_samples(batch, self.env.action_space.n)
+                            b_y = self._calc_q_update(b_ns, b_r, b_t)
+                            self._train_step(b_s, b_a, b_y)
+                            self.latest_metrics = self._calc_metrics(b_s, b_a, b_y)
+                            if (n_updates % self.target_update_interval == 0):
+                                self._sync_target_network()
+                            n_updates += 1
+
+                    if (n_samples % self.report_interval == 0):
+                        avg_reward = episode_reward/n_episodes if n_episodes != 0 else 0
+                        print("reward/episode since last report: %f" % avg_reward)
+                        print(self.metric_str % tuple(self.latest_metrics))
+                        print("%d experiences sampled" % n_samples)
+                        print("")
+                        saver.save(sess, "%s/model" % self.save_dir,
+                                   global_step=n_samples-self.num_burn_in)
```

Comment:
Dont think you are doing training with loaded weights correctly. If you load the weights of 10000 iterations, then it will overwrite the weight files 1000, 2000, 3000, ... with 11000, 12000, 13000, ...